docs: Add documentation for VolumeGroupSnapshot (#14720)
rook/rook: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-volumegroup-snapshot.md

---
title: VolumeGroup Snapshots
---

## Prerequisites

> - Install the [snapshot controller and snapshot v1 CRD](https://github.com/kubernetes-csi/external-snapshotter/tree/master#usage).

Reviewer: Pre req already mentioned in the kubernetes like you mentioned here https://github.com/kubernetes-csi/external-snapshotter/tree/master#volume-group-snapshot-support, just add a pointer to refer to the doc for all the pre-req

- Add the required args to the snapshot controller for volumegroup snapshot support. See [here](https://github.com/kubernetes-csi/external-snapshotter/tree/master#volume-group-snapshot-support) for more details.

> - We also need a `VolumeGroupSnapshotClass` for volumegroup snapshot to work. The purpose of a `VolumeGroupSnapshotClass` is

Reviewer: `volumegroup snapshot` to `volume group snapshot`

defined in [the kubernetes
documentation](https://kubernetes.io/blog/2023/05/08/kubernetes-1-27-volume-group-snapshot-alpha/).
In short, as the documentation describes it:

!!! info
    Created by cluster administrators to describe how volume group snapshots
    should be created. including the driver information, the deletion policy, etc.

## CephFS Snapshots

### VolumeGroupSnapshotClass

In [VolumeGroupSnapshotClass](https://github.com/rook/rook/tree/master/deploy/examples/csi/cephfs/groupsnapshotclass.yaml),
the `csi.storage.k8s.io/group-snapshotter-secret-name` parameter should reference the

> name of the secret created for the cephfsplugin.

Reviewer: `cephfsplugin` to `cephfs-plugin`

In the `VolumeGroupSnapshotClass`, update the value of the `clusterID` field to match the namespace
that Rook is running in. When Ceph CSI is deployed by Rook, the operator will automatically
maintain a configmap whose contents will match this key. By default this is
"rook-ceph".

```console
kubectl create -f deploy/examples/csi/cephfs/groupsnapshotclass.yaml
```

> ### VolumeGroupSnapshot

Reviewer: ```suggestion
### CephFS VolumeGroupSnapshot
```

In [VolumeGroupSnapshot](https://github.com/rook/rook/tree/master/deploy/examples/csi/cephfs/groupsnapshot.yaml),
`volumeGroupSnapshotClassName` should be the name of the `VolumeGroupSnapshotClass`

> previously created. The labels inside `matchLabels` should be the present on the
> PVC which is already created by the CephFS CSI driver.

Reviewer: ```suggestion
previously created. The labels inside `matchLabels` should be present on the
PVCs that are already created by the CephFS CSI driver.
```

```console
kubectl create -f deploy/examples/csi/cephfs/groupsnapshot.yaml
```

### Verify CephFS GroupSnapshot Creation

```console
$ kubectl get volumegroupsnapshotclass
NAME                              DRIVER                          DELETIONPOLICY   AGE
csi-cephfsplugin-groupsnapclass   rook-ceph.cephfs.csi.ceph.com   Delete           21m
```

```console
$ kubectl get volumegroupsnapshot
NAME                       READYTOUSE   VOLUMEGROUPSNAPSHOTCLASS          VOLUMEGROUPSNAPSHOTCONTENT                              CREATIONTIME   AGE
cephfs-pvc-groupsnapshot   true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-77ef7c5b-93ae-4ec3-b880-fa70740fb681   2m27s          2m29s
```

The snapshot will be ready to restore to a new PVC when `READYTOUSE` field of the
`volumegroupsnapshot` is set to true.

### Restore the CephFS groupsnapshot to a new PVC

In
[pvc-group-restore](https://github.com/rook/rook/tree/master/deploy/examples/csi/cephfs/pvc-restoregroup.yaml),
`dataSource` should be the name of the `VolumeGroupSnapshot` previously

> created. The `dataSource` kind should be the `VolumeSnapshot`.

Reviewer: This seems to be wrong we need to first identify the volumesnapshot name(s) from the volumegroupsnapshot object and use that for restore operation.

Author: Right, will update :)

Create a new PVC from the snapshot

```console
kubectl create -f deploy/examples/csi/cephfs/pvc-restoregroup.yaml
```

### Verify CephFS Restore PVC Creation

```console
$ kubectl get pvc
NAME                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
cephfs-pvc                Bound    pvc-9ae60bf9-4931-4f9a-9de1-7f45f31fe4da   1Gi        RWO            rook-cephfs    <unset>                 26m
cephfs-pvc-restoregroup   Bound    pvc-9ae60bf9-4931-4f9a-9de1-7f45f31fe4da   1Gi        RW0            rook-cephfs    <unset>                 30s
```

> ## CephFS snapshot resource Cleanup

Reviewer: ```suggestion
## CephFS volume group snapshot resource Cleanup
```

To clean your cluster of the resources created by this example, run the following:

```console
kubectl delete -f deploy/examples/csi/cephfs/pvc-restoregroup.yaml
kubectl delete -f deploy/examples/csi/cephfs/groupsnapshot.yaml
kubectl delete -f deploy/examples/csi/cephfs/groupsnapshotclass.yaml
```

## Limitations

- There is a limit of 400 snapshots per cephFS filesystem.

> - The PVC cannot be deleted if it has snapshots. make sure all the snapshots on the PVC are deleted before you delete the PVC.

Reviewer: we dont have this limitation any more, we can remove this from here and from snapshot documentation as well

---
title: VolumeGroup Snapshots
---

> 

Reviewer 2: Can you provide an intro paragraph to describe what volumegroup snapshots are, and the main scenario(s) where they are useful?

## Prerequisites

- Install the [snapshot controller and snapshot v1 CRD](https://github.com/kubernetes-csi/external-snapshotter/tree/master#usage),
refer to VolumeGroupSnapshot documentation
[here](https://github.com/kubernetes-csi/external-snapshotter/tree/master#volume-group-snapshot-support) for more details.

> - We also need a `VolumeGroupSnapshotClass` for volume group snapshot to work. The purpose of a `VolumeGroupSnapshotClass` is

Reviewer 2: ```suggestion
- A `VolumeGroupSnapshotClass` is needed for the volume group snapshot. The purpose of a `VolumeGroupSnapshotClass` is
```

defined in [the kubernetes
documentation](https://kubernetes.io/blog/2023/05/08/kubernetes-1-27-volume-group-snapshot-alpha/).

> In short, as the documentation describes it:

Reviewer 2: I'd suggest deleting lines 14-18. I don't understand them. They can just naturally keep reading this doc to learn more.

Reviewer: @travisn yes it can be deleted but we have similar one in the volume snapshot doc as well, This Doc is trying to align with the current format we are following.

Reviewer 2: hmm, maybe it needs to be reworded, I just don't understand it.

`volumeGroupSnapshotClassName` should be the name of the `VolumeGroupSnapshotClass`
previously created. The labels inside `matchLabels` should be present on the
PVCs that are already created by the CephFS CSI driver.

```console
kubectl create -f deploy/examples/csi/cephfs/groupsnapshot.yaml
```

### Verify CephFS GroupSnapshot Creation

```console
$ kubectl get volumegroupsnapshotclass
NAME                              DRIVER                          DELETIONPOLICY   AGE
csi-cephfsplugin-groupsnapclass   rook-ceph.cephfs.csi.ceph.com   Delete           21m
```

```console
$ kubectl get volumegroupsnapshot
NAME                       READYTOUSE   VOLUMEGROUPSNAPSHOTCLASS          VOLUMEGROUPSNAPSHOTCONTENT                              CREATIONTIME   AGE
cephfs-pvc-groupsnapshot   true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-d13f4d95-8822-4729-9586-4f222a3f788e   5m37s          5m39s
```

The snapshot will be ready to restore to a new PVC when `READYTOUSE` field of the
`volumegroupsnapshot` is set to true.

### Restore the CephFS groupsnapshot to a new PVC

> We need to find the name of the snapshots created by the `VolumeGroupSnapshot` first by running:

Reviewer 2: ```suggestion
Find the name of the snapshots created by the `VolumeGroupSnapshot` by running:
```

```console

> $ kubectl get volumegroupsnapshot/cephfs-pvc-groupsnapshot -o=jsonpath='{range .status.pvcVolumeSnapshotRefList[*]}PVC: {.persistentVolumeClaimRef.name}, Snapshot: {.volumeSnapshotRef.name}{"\n"}{end}'

Reviewer: please add namespace flag to the CLI

Author: We do not use explicit namespace in any of the other resources.

PVC: cephfs-pvc, Snapshot: snapshot-9d21b143904c10f49ddc92664a7e8fe93c23387d0a88549c14337484ebaf1011-2024-09-12-3.49.13
```

It will list the PVC's name followed by its snapshot name.

In
[pvc-restore](https://github.com/rook/rook/tree/master/deploy/examples/csi/cephfs/pvc-restore.yaml),
`dataSource` should be one of the `Snapshot` that we just
found. The `dataSource` kind should be the `VolumeSnapshot`.

Create a new PVC from the snapshot

```console
kubectl create -f deploy/examples/csi/cephfs/pvc-restore.yaml
```

### Verify CephFS Restore PVC Creation

```console
$ kubectl get pvc
cephfs-pvc           Bound    pvc-9ae60bf9-4931-4f9a-9de1-7f45f31fe4da   1Gi        RWO            rook-cephfs    <unset>                 171m
cephfs-pvc-restore   Bound    pvc-b4b73cbb-5061-48c7-9ac8-e1202508cf97   1Gi        RWO            rook-cephfs    <unset>                 46s
```

## CephFS volume group snapshot resource Cleanup

> To clean your cluster of the resources created by this example, run the following:

Reviewer 2: ```suggestion
To clean the resources created by this example, run the following:
```

```console
kubectl delete -f deploy/examples/csi/cephfs/pvc-restore.yaml
kubectl delete -f deploy/examples/csi/cephfs/groupsnapshot.yaml
kubectl delete -f deploy/examples/csi/cephfs/groupsnapshotclass.yaml
```

## Limitations

> - The PVC cannot be deleted if it has snapshots. make sure all the snapshots on the PVC are deleted before you delete the PVC.

Reviewer 2: ```suggestion
- The PVC cannot be deleted if it has snapshots. All snapshots on the PVC must be deleted before the PVC is deleted.
```

Reviewer: we should remove this limitation from both group snapshot and snapshot, we have the flexibility to remove snapshot after the clone is done

---
title: VolumeGroup Snapshots
---

> A volume group snapshot represents “copies” from multiple volumes that are taken at the same point-in-time.
> It can be used either to rehydrate new volumes (pre-populated with the snapshot data) or
> to restore existing volumes to a previous state (represented by the snapshots).

Reviewer: ```suggestion
Ceph provides the ability to create a crash-consistent snapshot of multiple volumes. A group snapshot represents “copies” from multiple volumes that are taken at the same point in time. A group snapshot can be used either to rehydrate new volumes (pre-populated with the snapshot data) or to restore existing volumes to a previous state (represented by the snapshots)
```

## Prerequisites

> - Install the [snapshot controller and snapshot v1 CRD](https://github.com/kubernetes-csi/external-snapshotter/tree/master#usage),

Reviewer: `snapshot controller and snapshot v1 CRD` to `snapshot controller and snapshot and volume group snapshot CRD`

---

> title: VolumeGroup Snapshots

Reviewer 3: as it is cephfs specific can we rename file with cephfs keyword

Author: The RBD docs will be added later in the same file. Hence the agnostic naming.

Reviewer: We have plan to update this doc for rbd as well.

Reviewer: `VolumeGroup Snapshots` to `Volume Group Snapshots` please change volumegroup to `volume-group` in file name as well

---

Ceph provides the ability to create crash-consistent snapshots of multiple volumes.
A group snapshot represents “copies” from multiple volumes that are taken at the same point in time.
A group snapshot can be used either to rehydrate new volumes (pre-populated with the snapshot data)
or to restore existing volumes to a previous state (represented by the snapshots)

## Prerequisites

- Install the [snapshot controller, volume group snapshot and snapshot CRDs](https://github.com/kubernetes-csi/external-snapshotter/tree/master#usage),
refer to VolumeGroupSnapshot documentation
[here](https://github.com/kubernetes-csi/external-snapshotter/tree/master#volume-group-snapshot-support) for more details.

- A `VolumeGroupSnapshotClass` is needed for the volume group snapshot to work. The purpose of a `VolumeGroupSnapshotClass` is
defined in [the kubernetes
documentation](https://kubernetes.io/blog/2023/05/08/kubernetes-1-27-volume-group-snapshot-alpha/).
In short, as the documentation describes it:

!!! info
    Created by cluster administrators to describe how volume group snapshots
    should be created. including the driver information, the deletion policy, etc.

> ## CephFS Snapshots

Reviewer: `## CephFS Snapshots` to `## Volume Group Snapshots`

> ### VolumeGroupSnapshotClass

Reviewer: `### VolumeGroupSnapshotClass` to `### CephFS VolumeGroupSnapshotClass`

```

### CephFS VolumeGroupSnapshot

In [VolumeGroupSnapshot](https://github.com/rook/rook/tree/master/deploy/examples/csi/cephfs/groupsnapshot.yaml),
`volumeGroupSnapshotClassName` should be the name of the `VolumeGroupSnapshotClass`
previously created. The labels inside `matchLabels` should be present on the
PVCs that are already created by the CephFS CSI driver.

```console
kubectl create -f deploy/examples/csi/cephfs/groupsnapshot.yaml
```

### Verify CephFS GroupSnapshot Creation

```console
$ kubectl get volumegroupsnapshotclass
NAME                              DRIVER                          DELETIONPOLICY   AGE
csi-cephfsplugin-groupsnapclass   rook-ceph.cephfs.csi.ceph.com   Delete           21m
```

```console
$ kubectl get volumegroupsnapshot
NAME                       READYTOUSE   VOLUMEGROUPSNAPSHOTCLASS          VOLUMEGROUPSNAPSHOTCONTENT                              CREATIONTIME   AGE
cephfs-pvc-groupsnapshot   true         csi-cephfsplugin-groupsnapclass   groupsnapcontent-d13f4d95-8822-4729-9586-4f222a3f788e   5m37s          5m39s
```

The snapshot will be ready to restore to a new PVC when `READYTOUSE` field of the
`volumegroupsnapshot` is set to true.

> ### Restore the CephFS groupsnapshot to a new PVC

Reviewer: `groupsnapshot` to `volume group snapshot`

Find the name of the snapshots created by the `VolumeGroupSnapshot` first by running:

```console

> $ kubectl get volumegroupsnapshot/cephfs-pvc-groupsnapshot -o=jsonpath='{range .status.pvcVolumeSnapshotRefList[*]}PVC: {.persistentVolumeClaimRef.name}, Snapshot: {.volumeSnapshotRef.name}{"\n"}{end}'

Reviewer: `cephfs-pvc-groupsnapshot` to `cephfs-groupsnapshot`
